File: scripts/units.cpp
/*
 * Compilation: g++ -o units units.cpp
 * Execution: ./units
 */

#include <iostream>
#include <boost/units/systems/si.hpp>

using namespace boost::units;

quantity<si::velocity> calculateSpeed(quantity<si::length> distance, quantity<si::time> duration)
{
    return distance / duration;
}

int main()
{
    double distance = 12, duration = 4;
    quantity<si::velocity> value = calculateSpeed(distance * si::meters, duration * si::seconds);
    std::cout << "Speed: " << value.value() << " m/s" << std::endl;
    return 0;
}

File: scripts/units.ml
let distance : <m> dfloat = create 12. and time : <s> dfloat = create 4.
let speed distance time = distance /: time
(* val speed: <m> Dim.dfloat -> <s> Dim.dfloat -> <m / s> Dim.dfloat = fun *)

let foot : <m/ft> dfloat = create 0.3048
let feet_to_meters (x : <ft> dfloat) = x *: foot
